src/main/java/io/jenkins/plugins/pipelinegraphview/treescanner/PipelineNodeTreeScanner.java

```diff
@@ -0,0 +1,424 @@
+                    wrappedNode.getTiming(),
+                    wrappedNode.getInputStep(),
+                    wrappedNode.getRun(),
+                    wrappedNode.getType());
+            FlowNodeWrapper closestParent = findParentNode(wrappedNode, stageMap);
+            logger.debug(String.format(
+                    "Found closest parent for node %s, %s",
+                    wrappedNode.getId(), (closestParent != null) ? closestParent.getId() : "null"));
+            if (closestParent != null) {
+                remappedNode.addParent(closestParent);
+                remappedNode.addEdge(closestParent);
+            }
+            return remappedNode;
+        }
+
+        /*
+         * This will remove any Stages and Step Blocks from the mapping. Still will be
+         * children of stages.
+         * This provides a similar representation to PipelineStepVisitor.
+         */
+        public @NonNull Map<String, FlowNodeWrapper> getStepMapping() {
+            if (wrappedStepMap != null) {
+                return wrappedStepMap;
+            }
+            Map<String, FlowNodeWrapper> stepMap = this.wrappedNodeMap.entrySet().stream()
+                    .filter(e -> (e.getValue().isStep()
+                            || e.getValue().getType() == FlowNodeWrapper.NodeType.UNHANDLED_EXCEPTION))
+                    .collect(Collectors.toMap(e -> e.getKey(), e -> e.getValue()));
+
+            Map<String, FlowNodeWrapper> stageMap = this.getStageMapping();
+            List<FlowNodeWrapper> remappedSteps = new ArrayList<>();
+            for (FlowNodeWrapper step : stepMap.values()) {
+                FlowNodeWrapper firstParent = step.getFirstParent();
+                // Remap parentage of steps that aren't children of stages (e.g. are in Step
+                // Block).
+                if (firstParent != null && !stageMap.containsKey(firstParent.getId())) {
+                    // Add to separate map to avoid updating the original map in loop.
+                    remappedSteps.add(remapNode(step, stageMap));
+                }
+            }
+            for (FlowNodeWrapper remappedStep : remappedSteps) {
+                stepMap.put(remappedStep.getId(), remappedStep);
+            }
+
+            this.wrappedStepMap = stepMap;
+            return this.wrappedStepMap;
+        }
+
+        /*
+         * Builds a graph from the list of nodes and relationships given to the class.
+         */
+        private void buildGraph() {
+            // ArrayDeque<FlowNodeWrapper> startNodes = new ArrayDeque<>();
+            List<FlowNode> nodeList = new ArrayList<FlowNode>(nodeMap.values());
+            Collections.sort(nodeList, new FlowNodeWrapper.FlowNodeComparator());
+            // If the Pipeline ended with an unhandled exception, then we want to catch the
+            // node which threw it.
+            BlockEndNode<?> nodeThatThrewException = getUnhandledException(nodeList.get(nodeList.size() - 1));
```

Comment:
I haven't quite figured out how to reproduce but twice I've managed to get:

```
2024-03-07 14:11:56.493+0000 [id=114]	WARNING	h.i.i.InstallUncaughtExceptionHandler#handleException: Caught unhandled exception with ID c30ceba8-e24a-4f8c-a4c1-a7f797ed6e6a
java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
	at java.base/java.util.Objects.checkIndex(Objects.java:372)
	at java.base/java.util.ArrayList.get(ArrayList.java:459)
	at io.jenkins.plugins.pipelinegraphview.treescanner.PipelineNodeTreeScanner$GraphBuilder.buildGraph(PipelineNodeTreeScanner.java:281)
	at io.jenkins.plugins.pipelinegraphview.treescanner.PipelineNodeTreeScanner$GraphBuilder.<init>(PipelineNodeTreeScanner.java:170)
	at io.jenkins.plugins.pipelinegraphview.treescanner.PipelineNodeTreeScanner.build(PipelineNodeTreeScanner.java:77)
```

so this needs a guard and can't just do size() - 1